File: problems/234. Palindrome Linked List/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        length = self.getCount(head)
        isEven = self.getCount(head) % 2 == 0
        mid = length // 2 if isEven else (length // 2) + 1
        pointer = head
        runner = head
        counter = 0
        
        while counter < mid:
            runner = runner.next
            counter += 1
        
        runner = self.reverseList(runner)
        
        while runner:
            if pointer.val != runner.val:
                return False
            pointer = pointer.next
            runner = runner.next
            
        return True
    
    def getCount(self, root):
        count = 0
        curr = root
        
        while curr:
            count += 1
            curr = curr.next
        
        return count
    
    def reverseList(self, head):
        if not head or not head.next:
            return head
        
        prev = None
        curr = head
        next_ = curr.next
        
        while curr.next:
            curr.next = prev
            
            prev = curr
            curr = next_
            next_ = curr.next
        curr.next = prev
        
        return curr
        

File: problems/234. Palindrome Linked List/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
    bool isPalindrome(ListNode *head)
    {
        ListNode *mid = head;
        ListNode *runner = head;

        while (runner)
        {
            mid = mid->next;
            if (!runner->next)
                break;
            runner = runner->next->next;
        }

        mid = reverse(mid);
        ListNode *left = head;
        ListNode *right = mid;

        while (right)
        {
            if (left->val != right->val)
                return false;
            left = left->next;
            right = right->next;
        }

        return true;
    }

    ListNode *reverse(ListNode *node)
    {
        if (!node || !node->next)
            return node;
        ListNode *p = reverse(node->next);
        node->next->next = node;
        node->next = NULL;
        return p;
    }
};